lib/puppet/application/device.rb

```diff
@@ -318,8 +318,37 @@ def main
             Puppet.info _("starting applying configuration to %{target} at %{scheme}%{url_host}%{port}%{url_path}") % { target: device.name, scheme: scheme, url_host: device_url.host, port: port, url_path: device_url.path }
             # this will reload and recompute default settings and create the devices sub vardir
             Puppet.settings.use :main, :agent, :ssl
-            # ask for a ssl cert if needed, but at least
-            # setup the ssl system for this device.
+
+            # Workaround for PUP-9642 / https://github.com/puppetlabs/puppet/pull/7483#issuecomment-483455997
+            # Josh Cooper wrote:
+            # > I'm reluctant to change this globally. Currently the puppet-agent package is responsible for setting the file mode of the
+            # > confdir during installation. If the package and puppet defaults differ, then it can cause the agent run to flap or conflict.
+            # >
+            # > Since the device application is the one creating the directory for a given device, I think it should enforce the desired mode,
+            # > and then let puppet manage the permissions for things inside the confdir.
+            # >
+            # > Also note to be consistent, the device application should change the owner and group of the device-specific confdir to the
+            # > Puppet[:user] and Puppet[:group] if they exist (which should only happen if the puppetserver package has been installed).
+            # > You can do something like:
+            # >
+            # > if Puppet.features.root?
+            # >   user = Puppet::Type.type(:user).new(name: Puppet[:user]).exists? ? Puppet[:user] : nil
+            # >   group = Puppet::Type.type(:group).new(name: Puppet[:group]).exists? Puppet[:group] : nil
+            # >   File.chown(user, group) if user || group
+            # > end
+
+            # This has to happen after `settings.use` above, so the directory is created and before `setup_host` below, where the SSL
+            # routines would fail with access errors
+            if Puppet.features.root?
```

Comment:
Should probably skip this on Windows too using `Puppet::Util::Platform.windows?`, otherwise if you run `puppet device` on Windows it will raise an exception due to `File.chown` not being implemented.

Reply:
good point